File: flutter_app/lib/models/relative.dart
class Relative {
  final String? relativeId;
  final String? firstName;
  final String? lastName;
  final DateTime? dateOfBirth;
  final String? gender;
  final String? contactAddress;
  final String? phoneNumber;
  final String? occupation;
  final String? email;
  final DateTime? createdDate;
  final DateTime? updatedDate;

  Relative({
    this.relativeId,
    this.firstName,
    this.lastName,
    this.dateOfBirth,
    this.gender,
    this.contactAddress,
    this.phoneNumber,
    this.occupation,
    this.email,
    this.createdDate,
    this.updatedDate,
  });

  factory Relative.fromJson(Map<String, dynamic> json) {
    return Relative(
      relativeId: json['relativeId'] as String?,
      firstName: json['firstName'] as String?,
      lastName: json['lastName'] as String?,
      dateOfBirth: json['dateOfBirth'] != null
          ? DateTime.tryParse(json['dateOfBirth'])
          : null,
      gender: json['gender'] as String?,
      contactAddress: json['contactAddress'] as String?,
      phoneNumber: json['phoneNumber'] as String?,
      occupation: json['occupation'] as String?,
      email: json['email'] as String?,
      createdDate: json['createdDate'] != null
          ? DateTime.tryParse(json['createdDate'])
          : null,
      updatedDate: json['updatedDate'] != null
          ? DateTime.tryParse(json['updatedDate'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'relativeId': relativeId,
      'firstName': firstName,
      'lastName': lastName,
      'dateOfBirth': dateOfBirth?.toIso8601String(),
      'gender': gender,
      'contactAddress': contactAddress,
      'phoneNumber': phoneNumber,
      'occupation': occupation,
      'email': email,
    };
  }

  Map<String, dynamic> toUpdateJson() {
    return {
      'phoneNumber': phoneNumber,
      'email': email,
      'contactAddress': contactAddress,
    };
  }

  String get fullName => '${firstName ?? ''} ${lastName ?? ''}'.trim();

  Relative copyWith({
    String? relativeId,
    String? firstName,
    String? lastName,
    DateTime? dateOfBirth,
    String? gender,
    String? contactAddress,
    String? phoneNumber,
    String? occupation,
    String? email,
  }) {
    return Relative(
      relativeId: relativeId ?? this.relativeId,
      firstName: firstName ?? this.firstName,
      lastName: lastName ?? this.lastName,
      dateOfBirth: dateOfBirth ?? this.dateOfBirth,
      gender: gender ?? this.gender,
      contactAddress: contactAddress ?? this.contactAddress,
      phoneNumber: phoneNumber ?? this.phoneNumber,
      occupation: occupation ?? this.occupation,
      email: email ?? this.email,
    );
  }
}

File: backend/src/main/java/edu/university/grademanagement/model/entity/Relative.java
package edu.university.grademanagement.model.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Relative Entity
 * Represents a relative (parent/guardian) of a student
 */
@Entity
@Table(name = "RELATIVES", schema = "GMS_ADMIN")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Relative {

    @Id
    @Column(name = "relative_id", length = 10)
    private String relativeId;

    @Column(name = "first_name", nullable = false, length = 50)
    private String firstName;

    @Column(name = "last_name", nullable = false, length = 50)
    private String lastName;

    @Column(name = "date_of_birth")
    private LocalDate dateOfBirth;

    @Column(name = "gender", length = 10)
    private String gender;

    @Column(name = "contact_address", length = 200)
    private String contactAddress;

    @Column(name = "phone_number", nullable = false, length = 20)
    private String phoneNumber;

    @Column(name = "occupation", length = 100)
    private String occupation;

    @Column(name = "email", length = 100)
    private String email;

    @Column(name = "created_date")
    private LocalDateTime createdDate;

    @Column(name = "updated_date")
    private LocalDateTime updatedDate;

    @PrePersist
    protected void onCreate() {
        createdDate = LocalDateTime.now();
        updatedDate = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedDate = LocalDateTime.now();
    }

    public String getFullName() {
        return firstName + " " + lastName;
    }
}

